File: src/main/java/com/esaulpaugh/abiv3/RLPItem.java
/*
   Copyright 2022 Evan Saulpaugh

   Licensed under the Apache License, Version 2.0 (the "License");
   you may not use this file except in compliance with the License.
   You may obtain a copy of the License at

       http://www.apache.org/licenses/LICENSE-2.0

   Unless required by applicable law or agreed to in writing, software
   distributed under the License is distributed on an "AS IS" BASIS,
   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
   See the License for the specific language governing permissions and
   limitations under the License.
*/
package com.esaulpaugh.abiv3;

import java.math.BigInteger;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Iterator;
import java.util.List;
import java.util.NoSuchElementException;

/**
 * An immutable view of a portion of a (possibly mutable) byte array containing RLP-encoded data, starting at {@code index}
 * (inclusive) and ending at {@code endIndex} (exclusive), representing a single item (either a string or list). Useful
 * when decoding or otherwise manipulating RLP items.
 *
 * Created by Evo on 1/19/2017.
 */
public final class RLPItem implements Iterable<RLPItem> {

    final byte[] buffer;
    public final int index;

    public final int dataIndex;
    public final int dataLength;
    public final int endIndex;

    RLPItem(byte[] buffer, int index, int dataIndex, int dataLength, int endIndex) {
        this.buffer = buffer;
        this.index = index;
        this.dataIndex = dataIndex;
        this.dataLength = dataLength;
        this.endIndex = endIndex;
    }

    public List<RLPItem> elements() {
        List<RLPItem> arrayList = new ArrayList<>();
        RLPItem next = nextElement(dataIndex);
        while (next != null) {
            arrayList.add(next);
            next = nextElement(next.endIndex);
        }
        return arrayList;
    }

    private RLPItem nextElement(int idx) {
        if(idx >= endIndex) return null;
        return wrap(buffer, idx, endIndex);
    }

    /**
     * Returns the payload portion of this item only, and not the prefix.
     *
     * @return  the data part of the encoding
     */
    public byte[] data() {
        return Arrays.copyOfRange(buffer, dataIndex, endIndex);
    }

    public int asInt() {
        return Integers.getInt(buffer, dataIndex, dataLength);
    }

    public BigInteger asBigInt() {
        return Integers.getBigInt(buffer, dataIndex, dataLength);
    }

    public BigInteger asBigIntSigned() {
        return new BigInteger(data());
    }

    public static RLPItem wrap(byte[] buffer, int index, int containerEnd) {
        final byte lead = buffer[index];
        if(lead == V3.VERSION_IDENTIFIER && containerEnd == index + 1) {
            return null;
        }
        final DataType type = DataType.type(lead);
        switch (type) {
        case SINGLE_BYTE: return newSingleByte(buffer, index, containerEnd);
        case STRING_SHORT: return newStringShort(buffer, index, lead, containerEnd);
        case LIST_SHORT: return newListShort(buffer, index, lead, containerEnd);
        case STRING_LONG:
        case LIST_LONG: return newLongItem(lead, type, buffer, index, containerEnd);
        default: throw new AssertionError();
        }
    }

    private static RLPItem newSingleByte(byte[] buffer, int index, int containerEnd) {
        final int endIndex = requireInBounds(index + 1L, containerEnd, index);
        return new RLPItem(buffer, index, index, 1, endIndex);
    }

    private static RLPItem newStringShort(byte[] buffer, int index, byte lead, int containerEnd) {
        final int dataIndex = index + 1;
        final int dataLength = lead - DataType.STRING_SHORT.offset;
        final int endIndex = requireInBounds((long) dataIndex + dataLength, containerEnd, index);
        if (dataLength == 1 && DataType.type(buffer[dataIndex]) == DataType.SINGLE_BYTE) {
            throw new IllegalArgumentException("invalid rlp for single byte @ " + index);
        }
        return new RLPItem(buffer, index, dataIndex, dataLength, endIndex);
    }

    private static RLPItem newListShort(byte[] buffer, int index, byte lead, int containerEnd) {
        final int dataIndex = index + 1;
        final int dataLength = lead - DataType.LIST_SHORT.offset;
        final int endIndex = requireInBounds((long) dataIndex + dataLength, containerEnd, index);
        return new RLPItem(buffer, index, dataIndex, dataLength, endIndex);
    }

    private static RLPItem newLongItem(byte lead, DataType type, byte[] buffer, int index, int containerEnd) {
        final int diff = lead - type.offset;
        final int lengthIndex = index + 1;
        final int dataIndex = requireInBounds((long) lengthIndex + diff, containerEnd, index);
        final long dataLength = Integers.getLong(buffer, lengthIndex, diff);
        if(dataLength < DataType.MIN_LONG_DATA_LEN) {
            throw new IllegalArgumentException("long element data length must be " + DataType.MIN_LONG_DATA_LEN
                    + " or greater; found: " + dataLength + " for element @ " + index);
        }
        final int dataLen = requireInBounds(dataLength, containerEnd, index);
        final int endIndex = requireInBounds(dataIndex + dataLength, containerEnd, index);
        return new RLPItem(buffer, index, dataIndex, dataLen, endIndex);
    }

    private static int requireInBounds(long val, int containerEnd, int index) {
        if (val > containerEnd) {
            String msg = "element @ index " + index + " exceeds its container: " + val + " > " + containerEnd;
            throw new IllegalArgumentException(msg);
        }
        return (int) val;
    }

    @Override
    public String toString() {
        byte[] rlp = new byte[endIndex - index];
        System.arraycopy(buffer, index, rlp, 0, rlp.length);
        return new BigInteger(1, rlp).toString(16);
    }

    @Override
    public Iterator<RLPItem> iterator() {
        return new ABIv3Iterator(buffer, dataIndex, endIndex);
    }

    static final class ABIv3Iterator implements Iterator<RLPItem> {

        final byte[] buffer;
        int index;
        final int containerEnd;

        RLPItem next;

        private ABIv3Iterator(byte[] buffer, int index, int containerEnd) {
            this.buffer = buffer;
            this.index = index;
            this.containerEnd = containerEnd;
        }

        @Override
        public boolean hasNext() {
            if (next != null) {
                return true;
            }
            if (index < containerEnd) {
                next = wrap(buffer, index, containerEnd);
                if(next == null) {
                    return false;
                }
                this.index = next.endIndex;
                return true;
            }
            return false;
        }

        @Override
        public RLPItem next() {
            if(hasNext()) {
                RLPItem item = next;
                next = null;
                index = item.endIndex;
                return item;
            }
            throw new NoSuchElementException();
        }

        static ABIv3Iterator sequenceIterator(byte[] buffer, int index) {
            return new ABIv3Iterator(buffer, index, buffer.length);
        }
    }
}

File: python/abiv3/RLPItem.py
# Copyright 2022 Evan Saulpaugh
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
from abiv3 import Utils
from abiv3.RLPIterator import RLPIterator


class RLPItem:

    def __init__(self, buffer, index, data_index, data_length, end_index):
        self.buffer = buffer
        self.index = index
        self.dataIndex = data_index
        self.dataLength = data_length
        self.endIndex = end_index

    def data(self):
        return self.buffer[self.dataIndex: self.endIndex: 1]

    def as_int(self):
        return int.from_bytes(self.data(), byteorder='big')

    def as_int_signed(self):
        return int.from_bytes(self.data(), byteorder='big', signed=True)

    def elements(self):
        the_list = []
        next_item = self.next_element(self.dataIndex)
        while next_item is not None:
            the_list.append(next_item)
            next_item = self.next_element(next_item.endIndex)
        return the_list

    def next_element(self, idx):
        if idx >= self.endIndex:
            return None
        return Utils.wrap(self.buffer, idx, self.endIndex)

    def iterator(self):
        return RLPIterator(self.buffer, self.dataIndex, self.endIndex)

    @staticmethod
    def rlp_type(lead):
        lead = Utils.to_signed_byte(lead)
        if lead < -72:  # 0xB8
            return 1  # short string
        if lead < -64:  # 0xC0
            return 2  # long string
        if lead < -8:  # 0xF8
            return 3  # short list
        if lead < 0:
            return 4  # long list
        return 0  # single byte

    @staticmethod
    def require_in_bounds(val, container_end, index):
        if val > container_end:
            msg = f'element @ index {index} exceeds its container: {val} > {container_end}'
            raise Exception(msg)
        return val

    @staticmethod
    def single_byte(buffer, index, container_end):
        end_index = RLPItem.require_in_bounds(index + 1, container_end, index)
        return RLPItem(buffer, index, index, 1, end_index)

    @staticmethod
    def short_string(buffer, index, lead, container_end):
        data_index = index + 1
        data_length = Utils.to_signed_byte(lead) - Utils.to_signed_byte(0x80)
        end_index = RLPItem.require_in_bounds(data_index + data_length, container_end, index)
        if data_length == 1 and RLPItem.rlp_type(buffer[data_index] == 0):
            raise Exception('invalid rlp for single byte @ ' + index)
        return RLPItem(buffer, index, data_index, data_length, end_index)

    @staticmethod
    def short_list(buffer, index, lead, container_end):
        data_index = index + 1
        data_length = Utils.to_signed_byte(lead) - Utils.to_signed_byte(0xC0)
        end_index = RLPItem.require_in_bounds(data_index + data_length, container_end, index)
        return RLPItem(buffer, index, data_index, data_length, end_index)

    @staticmethod
    def long_item(lead, offset, buffer, index, container_end):
        diff = Utils.to_signed_byte(lead) - Utils.to_signed_byte(offset)
        length_index = index + 1
        data_index = RLPItem.require_in_bounds(length_index + diff, container_end, index)
        data_length = int.from_bytes(buffer[length_index: length_index + diff: 1], "big")
        if data_length < 56:
            raise Exception(f'long element data length must be 56 or greater; found: {data_length} for element @ {index}')
        data_len = RLPItem.require_in_bounds(data_length, container_end, index)
        end_index = RLPItem.require_in_bounds(data_index + data_length, container_end, index)
        return RLPItem(buffer, index, data_index, data_len, end_index)
